[WIP] build: Add CMake-based build system

hebasto opened the pull request:
**Benefits** of using CMake in the Bitcoin Core project:
- no hacks required to build `dll` for Windows, even with `DEBUG=1` (see https://github.com/bitcoin/bitcoin/issues/19772)
- better maintainability (say _bye_ to global variables)
- easy integration with Qt 5 (compare [`src/qt/CMakeLists.txt`](https://github.com/hebasto/bitcoin/blob/cmake/pr/src/qt/CMakeLists.txt) vs [`build-aux/m4/bitcoin_qt.m4`](https://github.com/hebasto/bitcoin/blob/cmake/pr/build-aux/m4/bitcoin_qt.m4))
- easy integration with [Qt 6](https://github.com/hebasto/bitcoin/pull/3#issuecomment-1153251584)
- works fine with installed both Qt 5 and Qt 6 on Linux and macOS

**Important notes**:
- backward compatibility with Ubuntu Bionic (only CMake 3.10 is required)
- see a new CI task :)
- building in the source tree is [not supported](https://github.com/hebasto/bitcoin/pull/3#issuecomment-1157499334).

Done:
  - [X] `bitcoind` executable, could be disabled with `-DBUILD_DAEMON=OFF`
  - [X] `bitcoin-cli` executable, could be disabled with `-DBUILD_CLI=OFF`
  - [X] `libbitcoinconsensus` shared library, could be disabled with `-DBUILD_BITCOINCONSENSUS_LIB=OFF`
  - [X] `libbitcoinkernel` shared library, could be disabled with `-DBUILD_BITCOINKERNEL_LIB=OFF`
  - [X] wallet support, could be disabled with `-DENABLE_WALLET=OFF`
  - [X] `bitcoin-qt` executable, could be specified with `-DWITH_GUI=Qt5` or `-DWITH_GUI=Qt6`, or disabled with `-DWITH_GUI=no` (no Android support)
  - [X] `ccache` support, could be disabled with `-DUSE_CCACHE=OFF` (not working when cross-compiling for macOS, see bitcoin/bitcoin#21552 and bitcoin/bitcoin#24620)
  - [X] cross-compiling, including Windows builds with `DEBUG=1`
  - [X] reduce exported symbols support, could be enabled with `-DREDUCE_EXPORTS=ON`
  - [X] optional packages support

Not implemented yet:
  - [ ] `bitcoin-qt` for Android
  - [ ] ...

Need to be fixed:
  - [ ] ...

---

### Native compiling has been tested on the following OSes:
- Ubuntu 22.04
- macOS Monterey
- FreeBSD 12.3
- OpenBSD 7.1
```sh
rm -rf build && cmake -S . -B build
cmake --build build
```

### Native compiling on Ubuntu 22.04 with system Qt 6
```sh
sudo apt install qt6-base-dev qt6-l10n-tools qt6-tools-dev qt6-tools-dev-tools libxkbcommon-dev
rm -rf build && cmake -S . -B build
cmake --build build
```

### Native compiling on macOS Monterey with Homebrew's Qt 6
```sh
brew install qt
rm -rf build && cmake -S . -B build
cmake --build build
```

### Cross compiling for Windows on Ubuntu 22.04
```sh
make -C depends HOST=x86_64-w64-mingw32
rm -rf build && cmake --toolchain depends/x86_64-w64-mingw32/share/toolchain.cmake -S . -B build
cmake --build build
```

### Cross compiling for macOS on Ubuntu 22.04
```sh
make -C depends HOST=x86_64-apple-darwin
rm -rf build && cmake --toolchain depends/x86_64-apple-darwin/share/toolchain.cmake -S . -B build
cmake --build build
```

---
## Configuration Options Comparison Table (a draft)
| Autotools | CMake |
|---|---|
| `--enable-ccache` |  `-DUSE_CCACHE=auto,yes,no` |

Optional packages:
| Autotools | CMake |
|---|---|
| `--with-gui` | `-DWITH_GUI=auto,Qt5,Qt6,no` |
| `--with-natpmp` | `-DWITH_NATPMP=auto,yes,no` |
| `--enable-natpmp-default` | `-DENABLE_NATPMP_DEFAULT=ON,OFF` |
| `--with-miniupnpc` | `-DWITH_MINIUPNPC=auto,yes,no` |
| `--enable-upnp-default` | `-DENABLE_UPNP_DEFAULT=ON,OFF` |
| `--enable-zmq` |  `-DWITH_ZMQ=auto,yes,no` |
| `--enable-usdt` |  `-DWITH_USDT=auto,yes,no` |
| `--with-seccomp` |  `-DWITH_SECCOMP=auto,yes,no` |
| `--with-qrencode` |  `-DWITH_QRENCODE=auto,yes,no` |

---

## Functional Tests

Functional tests can be run in exactly the same way as when building with Autotools out of the source tree, i.e.:
```
./build/test/functional/test_runner.py
```

---

A related brainstorming issue: https://github.com/hebasto/bitcoin/issues/4.

hebasto commented:
Added `config/bitcoin-config.h` support.

Fixed macro definitions for macOS.

hebasto commented on depends/Makefile:
```diff
@@ -246,6 +247,19 @@ $(host_prefix)/share/config.site : config.site.in $(host_prefix)/.stamp_$(final_
+$(host_prefix)/share/depends.cmake : depends.cmake.in $(host_prefix)/.stamp_$(final_build_id)
+	@mkdir -p $(@D)
+	sed -e 's|@CONFIG_SITE@|$(host_prefix)/share/config.site|' \
+        -e 's|@CMAKE_SYSTEM_NAME@|$($(host_os)_cmake_system)|' \
```
Yes, docs are clear.

FWIW, `CMAKE_SYSTEM_VERSION` is ignored currently: https://github.com/hebasto/bitcoin/blob/455780b1aebae21ebd5d83d2f0ff6a2264ac9608/depends/funcs.mk#L183-L185

ryanofsky replied:
> Yes, docs are clear.

I think this doc is overgeneralizing. If you look at https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html you can see this variable only has meaning when cross compiling for windows and android

fanquake replied:
So it might be Android and Windows, or Android and macOS, or Android and Windows and macOS? In either case it seems like we should figure out if we need to set this.

hebasto commented:
Added an initial implementation of `qt/CMakeList.txt`.

Addressed @fanquake's [comment](https://github.com/hebasto/bitcoin/pull/3#discussion_r892222440).

hebasto commented on CMakeLists.txt:
```diff
@@ -0,0 +1,141 @@
+
+# Going to be compatible with Qt 6.2 LTS.
+# See https://doc.qt.io/qt-6.2/cmake-get-started.html
+cmake_minimum_required(VERSION 3.16)
```
Thanks! [Updated](https://github.com/hebasto/bitcoin/pull/3#issuecomment-1151692850).

theStack reviewed (commented):
Tested building f217df3fb506751cf8f6896bc6b2fcb7094aed0e on OpenBSD 7.1 via
```
rm -rf build && cmake -DCMAKE_MAKE_PROGRAM=gmake -S . -B build
cmake --build build
```
The good news: the previously (via IRC) reported secp256k1 preprocessor failure is gone now and I could successfully build `bitcoin-cli` and `bitcoind` 🚀 however, the following patch was needed (could have probably also specified that as argument?):
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 94c2934d7..6fff6093d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -80,7 +80,7 @@ option(BUILD_CLI "Build bitcoin-cli" ON)
 option(BUILD_BITCOINCONSENSUS_LIB "Build bitcoinconsensus shared library" ON)
 option(BUILD_BITCOINKERNEL_LIB "Build experimental bitcoinkernel shared library" ON)
 option(ENABLE_WALLET "Enable wallet" ON)
-option(BUILD_QT5_GUI "Build Qt5 widget-based GUI" ON)
+option(BUILD_QT5_GUI "Build Qt5 widget-based GUI" OFF)

 include(cmake/subtree-crc32c.cmake)
 include(cmake/subtree-leveldb.cmake)
```

The bad news: building `bitcoin-qt` fails due to a missing CMake file (qt5 was installed via `# pkg_add qt5` which should bring the necessary development files):
```
CMake Error at src/qt/CMakeLists.txt:14 (find_package):
  By not providing "FindQt5.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Qt5", but
  CMake did not find one.

  Could not find a package configuration file provided by "Qt5" (requested
  version 5.11.3) with any of the following names:

    Qt5Config.cmake
    qt5-config.cmake

  Add the installation prefix of "Qt5" to CMAKE_PREFIX_PATH or set "Qt5_DIR"
  to a directory containing one of the above files.  If "Qt5" provides a
  separate development package or SDK, be sure it has been installed.


-- Configuring incomplete, errors occurred!
```
The file `Qt5Config.cmake` is available in `/usr/local/lib/qt5/cmake/Qt5/Qt5Config.cmake`, probably just some search path has to be extended here.

hebasto commented:
Reworked integration with depends.

`bitcoin-qt` can be cross compiled now (except for Android).

The build instructions in the PR description have been updated.

furszy commented:
Big concept ACK.

I'm having the same problem as @theStack in macOS Monterey.

This fixes it (tested on macOS only):

```cmake
if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
  list(APPEND CMAKE_PREFIX_PATH /usr/local/opt/qt5 /opt/homebrew/opt/qt5)
elseif(${CMAKE_SYSTEM_NAME} MATCHES "Linux")
  list(APPEND CMAKE_PREFIX_PATH /usr/lib/x86_64-linux-gnu/cmake/Qt5)
  set(Qt5core_DIR "/usr/lib/x86_64-linux-gnu/cmake/Qt5/QtCore")
endif()

find_package(Qt5 5.11.3 REQUIRED COMPONENTS Widgets LinguistTools)
```

hebasto commented:
Added support for system Qt6 on Linux.

hebasto commented:
@furszy 
> I'm having the same problem as @theStack in macOS Monterey.

Yes. It fails with `qt@5` (will be fixed).

But it builds successfully with `qt@6` from Homebrew :tiger2:

hebasto commented:
> I'm having the same problem as @theStack in macOS Monterey.

Fixed build with Homebrew's Qt 5 on macOS.

hebasto commented:
@Sjors 

Building GUI with wallet requires https://github.com/bitcoin/bitcoin/pull/25338.

hebasto commented:
@theStack 

> The bad news: building `bitcoin-qt` fails due to a missing CMake file (qt5 was installed via `# pkg_add qt5` which should bring the necessary development files):
> 
> ```
> CMake Error at src/qt/CMakeLists.txt:14 (find_package):
>   By not providing "FindQt5.cmake" in CMAKE_MODULE_PATH this project has
>   asked CMake to find a package configuration file provided by "Qt5", but
>   CMake did not find one.
> 
>   Could not find a package configuration file provided by "Qt5" (requested
>   version 5.11.3) with any of the following names:
> 
>     Qt5Config.cmake
>     qt5-config.cmake
> 
>   Add the installation prefix of "Qt5" to CMAKE_PREFIX_PATH or set "Qt5_DIR"
>   to a directory containing one of the above files.  If "Qt5" provides a
>   separate development package or SDK, be sure it has been installed.
> 
> 
> -- Configuring incomplete, errors occurred!
> ```
> 
> The file `Qt5Config.cmake` is available in `/usr/local/lib/qt5/cmake/Qt5/Qt5Config.cmake`, probably just some search path has to be extended here.

Additional options `-DBUILD_GUI=Qt5 -DQt5_DIR=/usr/local/lib/qt5/cmake/Qt5` are required. The example for OpenBSD 7.1 has been updated in the PR description.

Sjors commented:
GUI wallet works for me on macOS 12.4 with qt 6 now.

I get a few minor warnings (building without depends):
```
[ 38%] Linking CXX static library libbitcoin_util.a
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: libbitcoin_util.a(sync.cpp.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: libbitcoin_util.a(sync.cpp.o) has no symbols
```

There's also a deprecation warning ` warning: 'location' is deprecated: Use path() [-Wdeprecated-declarations]` but I'm guessing that's a QT 6 issue, not cmake.

It also seems to run the libsecp256k `./configure` during the  `cmake --build build` step, rather than the configuration step. This creates noise when compiling with `-jN`, because the configure output lines are mixed with building lines.

The two issues above make it difficult to notice any other warnings, so they might be worth addressing (and rebasing) for that reason.

I was also able to build and run with depends (without cross compilation), like so:

```sh
make -C depends
 rm -rf build && cmake --toolchain depends/x86_64-apple-darwin21.5.0/share/toolchain.cmake -S . -B build
```

The binaries now end up in build/src/bitcoind, so that either needs to change, or we need the functional tests to use the new location. There's also no `test/config.ini` yet, so I guess you're still working on that.

When building with distcc, I get bunch of `clang: warning: argument unused during compilation: '-iframework /usr/local/lib'`. This was without using depends, and the only other machine run the same macOS and XCode version, and has the same SDK.

hebasto commented:
Implemented backward compatibility with CMake 3.10 (Ubuntu Bionic).

Added a dedicated CI [task](https://cirrus-ci.com/task/5024588684853248).

hebasto commented:
@Sjors 
> And can we get it to print a similar summary?

Thanks! A draft implementation of the configure summary has been pushed.

luke-jr commented on cmake/subtree-leveldb.cmake:
```diff
@@ -0,0 +1,111 @@
+if(HAVE_FDATASYNC)
+  target_compile_definitions(leveldb PRIVATE HAVE_FDATASYNC=1)
+else()
+  target_compile_definitions(leveldb PRIVATE HAVE_FDATASYNC=0)
```
Do we actually want this? LevelDB's own CMake stuff acts on `HAVE_FDATASYNC` being *defined* rather than 1...

hebasto replied:
> LevelDB's own CMake stuff acts on `HAVE_FDATASYNC` being _defined_ rather than 1...

I don't think so.[ "LevelDB's own CMake stuff"](https://github.com/bitcoin-core/leveldb-subtree/blob/330dd6235f50c01e49b4cfadf6df9e026dc61f0d/port/port_config.h.in#L8-L11):
```
// Define to 1 if you have a definition for fdatasync() in <unistd.h>.
#if !defined(HAVE_FDATASYNC)
#cmakedefine01 HAVE_FDATASYNC
#endif  // !defined(HAVE_FDATASYNC)
```
_always_ defines the `HAVE_FDATASYNC` macro via [`#cmakedefine01`](https://cmake.org/cmake/help/latest/command/configure_file.html) replacement.

luke-jr reviewed (commented):
Why isn't this in the main repo?

(Inclined to concept NACK too, but ... our autotools is a mess anyway, sigh)

luke-jr commented on cmake/subtree-leveldb.cmake:
```diff
@@ -0,0 +1,111 @@
+if(HAVE_O_CLOEXEC)
+  target_compile_definitions(leveldb PRIVATE HAVE_O_CLOEXEC=1)
+else()
+  target_compile_definitions(leveldb PRIVATE HAVE_O_CLOEXEC=0)
```
This has issues with LevelDB as well (also a bug in master)

hebasto replied:
https://github.com/google/leveldb/pull/1028

0xB10C commented:
Thanks for working on this! I primarily took a look at the USDT tracepoints and if the work. They seem to work! I build on a`x84_64` machine running NixOS 22.05 with the same commands as you've listed for Ubuntu 22.04.

The binary `build/src/bitcoind` (new location) contains tracepoints (via [readelf](https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md#with-readelf)). To run the functional tests and the USDT interface tests (`test/functional/interface_usdt_*.py`) I copied a `test/config.ini` from another non-cmake build and appended `/build` to `BUILDDIR` and disabled e.g. `ENABLE_SYSCALL_SANDBOX`. The tracepoint interface tests pass. 

Ran the functional test suite and these tests failed for me. I suspect that some features aren't working but haven't looked into it closer.
```
feature_notifications.py                           | ✖ Failed  | 2 s
feature_startupnotify.py                           | ✖ Failed  | 3 s
feature_versionbits_warning.py                     | ✖ Failed  | 2 s
rpc_setban.py                                      | ✖ Failed  | 2 s
tool_wallet.py --descriptors                       | ✖ Failed  | 2 s
```

hebasto commented:
@luke-jr 
> Why isn't this in the main repo?

Discussed on [IRC](https://gnusha.org/bitcoin-builds/2022-06-18.log).

> Inclined to concept NACK

https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#conceptual-review:

> A NACK needs to include a rationale why the change is not worthwhile. NACKs without accompanying reasoning may be disregarded.

theStack commented:
> Additional options `-DBUILD_GUI=Qt5 -DQt5_DIR=/usr/local/lib/qt5/cmake/Qt5` are required. The example for OpenBSD 7.1 has been updated in the PR description.

Tested again on OpenBSD 7.1 with the following commands (as outlined in the updated PR description):
```
rm -rf build && cmake -DCMAKE_MAKE_PROGRAM=gmake -DBUILD_GUI=Qt5 -DQt5_DIR=/usr/local/lib/qt5/cmake/Qt5 -S . -B build
cmake --build build
```
Successfully built the binaries `bicoin-cli`, `bitcoind` and `bitcoin-qt` 👌

hebasto commented:
Added a draft implementation of the `test/config.ini` configuration file.

Functional tests:
- added to the CI task
- instructions added into the PR description

@Sjors 
> The binaries now end up in build/src/bitcoind, so that either needs to change, or we need the functional tests to use the new location. 

Actually, functional tests can be run in exactly the same way as when building with Autotools out of the source tree, i.e.:
```
./build/test/functional/test_runner.py
```

No changes required :)

hebasto commented:
@Sjors 
> I'm getting a bunch of:
> 
> ```
> In file included from /Users/sjors/dev/bitcoin-cmake/src/validation.h:28:
> /Users/sjors/dev/bitcoin-cmake/src/txmempool.h:28:10: fatal error: 'boost/multi_index/hashed_index.hpp' file not found
> #include <boost/multi_index/hashed_index.hpp>
> ```
> 
> boost: stable 1.78.0 (bottled)

Thanks for testing! Should be fixed now :)

hebasto commented:
@Sjors 

> There's also a deprecation warning ` warning: 'location' is deprecated: Use path() [-Wdeprecated-declarations]` but I'm guessing that's a QT 6 issue, not cmake.

Fixed in bitcoin-core/gui#623.

Sjors commented:
> Thanks for testing! Should be fixed now :)

Yup, that works. Test suite passed too on macOS. But how do I go about running an individual test? (`--filter` is a workaround)

External signer support doesn't get built yet.

hebasto commented:
> Yup, that works. Test suite passed too on macOS. But how do I go about running an individual test? (`--filter` is a workaround)

Tests code modifications (outside the build system) are required, therefore, leaving them for future follow ups.

hebasto commented:
`cmake/subtree-secp256k1.cmake` is pure CMake now :)

No using of Autotools to build secp256k1 library.

hebasto commented:
Removed `cmake/subtree-crc32c.cmake` file, and crc32c native CMake project is used.

Although, a bugfix is required. See: https://github.com/bitcoin-core/crc32c-subtree/pull/5.

hebasto commented:
Removed `cmake/subtree-leveldb.cmake` file, and leveldb native CMake project is used.

@Sjors 
> There's also a deprecation warning ` warning: 'location' is deprecated: Use path() [-Wdeprecated-declarations]` but I'm guessing that's a QT 6 issue, not cmake.

Should be fixed now :)

hebasto commented:
@w0xlt 

> Tested ACK [06b0ca6](https://github.com/hebasto/bitcoin/commit/06b0ca6d83d69fed713a81b431653b426636a030) on Ubuntu 21.10 via:
> 
> ```
> rm -rf build && cmake -S . -B build
> cmake --build build
> ```
> 
> I needed to install `libqrencode-dev`. After that, the above commands worked. I have run `bitcoind`, `bitcoin-qt` and functional tests successfully.

Thank you for testing!

`libqrencode-dev` is an _optional_ package. Its absence shouldn't end with an error.

hebasto commented:
@w0xlt 
> `libqrencode-dev` is an _optional_ package. Its absence shouldn't end with an error.

Should be fixed now.

w0xlt commented:
reACK https://github.com/hebasto/bitcoin/pull/3/commits/46a1f89248dee0ad423a476ae3a07f6104fc6cd7

Confirmed that `libqrencode-dev` is being considered optional when compiling on Ubuntu 21.10 and 22.04 .

```
-- Performing Test HAVE_USDT_H - Success
-- Checking for module 'libqrencode'
--   No package 'libqrencode' found
-- Performing Test HAVE_SECCOMP_H
```

hebasto commented:
- reworked / optimized dependencies handling
- BDB is supported now
- added one more task to CI

---

@fanquake 

Thank you for testing!

> When configuring from depends, CXXFLAGS are being passed to C code

Should be fixed now :tiger2:

hebasto commented:
- CI tasks are green now

@fanquake 

> After building with the wallet disabled, I see a number of functional tests failures that I don't see building master with the wallet disabled

Fixed. Thanks!

fanquake commented:
> reworked / optimized dependencies handling

Looks like this is still broken. On a macOS system with no Qt available:
```bash
rm -rf build && cmake -S . -B build
...
-- Performing Test HAVE_CXX17_HAS_INCLUDE
-- Performing Test HAVE_CXX17_HAS_INCLUDE - Success
CMake Error at src/qt/CMakeLists.txt:32 (find_package):
  Could not find a package configuration file provided by "QT" with any of
  the following names:

    Qt6Config.cmake
    qt6-config.cmake
    Qt5Config.cmake
    qt5-config.cmake

  Add the installation prefix of "QT" to CMAKE_PREFIX_PATH or set "QT_DIR" to
  a directory containing one of the above files.  If "QT" provides a separate
  development package or SDK, be sure it has been installed.


-- Configuring incomplete, errors occurred!
```

theStack commented:
> @theStack
> 
> > Additional options `-DBUILD_GUI=Qt5 -DQt5_DIR=/usr/local/lib/qt5/cmake/Qt5` are required.
> 
> No additional options are longer required when building on OpenBSD 7.1 🐅
> 
> `-DCMAKE_MAKE_PROGRAM=gmake` is no longer required as well.
> 
> The PR description has been updated.

That's great! Just restested building on OpenBSD 7.1 (commit c928933896c0731f9d38985fec7ee273a940a6f0), and unfortunately get an error regarding BerkeleyDB:
```
In file included from /home/honey/bitcoin_prrev/src/wallet/walletdb.cpp:18:
/home/honey/bitcoin_prrev/src/wallet/bdb.h:27:10: fatal error: 'db_cxx.h' file not found
#include <db_cxx.h>
         ^~~~~~~~~~
1 error generated.
*** Error 1 in . (src/wallet/CMakeFiles/bitcoin_wallet.dir/build.make:426 'src/wallet/CMakeFiles/bitcoin_wallet.dir/walletdb.cpp.o': cd /hom...)
*** Error 2 in . (CMakeFiles/Makefile2:762 'src/wallet/CMakeFiles/bitcoin_wallet.dir/all': /usr/bin/make -s -f src/wallet/CMakeFiles/bitcoin...)
*** Error 2 in /home/honey/bitcoin_prrev/build (Makefile:91 'all': /usr/bin/make -s -f CMakeFiles/Makefile2 all)
```
The installed BDB version is 4.6.21, installed via ports (`doas pkg_add db`). The detection in the beginning said:
```
-- Found BerkeleyDB: /usr/include (Required is at least version "4.8")
CMake Warning at cmake/optional.cmake:68 (message):
  Found Berkeley DB (BDB) other than 4.8.
Call Stack (most recent call first):
  CMakeLists.txt:99 (include)


CMake Warning at cmake/optional.cmake:72 (message):
  BDB (legacy) wallets opened by this build would not be portable!

  If this is intended, pass "-DALLOW_INCOMPATIBLE_BDB=ON".

  Passing "-DWITH_BDB=no" will suppress this warning.

Call Stack (most recent call first):
  CMakeLists.txt:99 (include)
```
Seems like the version installed is even lower than the required minimum (4.8), does it still make sense to allow incompatible then? (Totally not sure how I was even able to compile in the past by the way...)

By the way, is there any way to pass options to `(g)make`? Asking this primarily to activate parallelization for builds via `-j<N>`.

hebasto commented:
Rebased.
Addressed:
- @fanquake's https://github.com/hebasto/bitcoin/pull/3#issuecomment-1167072796
- @theStack's https://github.com/hebasto/bitcoin/pull/3#issuecomment-1175647461


@theStack 
> By the way, is there any way to pass options to `(g)make`? Asking this primarily to activate parallelization for builds via `-j<N>`.

```
cmake --build build -- -jN
```
or you can use `(g)make` directly.

hebasto commented:
Rebased.

Fixed some bugs.

Added `bitcoin-wallet` binary.

hebasto commented:
Rebased.

Added `bitcoin-tx` and `bitcoin-util` binaries.

hebasto commented:
https://github.com/bitcoin/bitcoin/pull/25797 has been opened without Qt 6 integration and CMake-specific CI tasks.

hebasto commented:
Synced with the bitcoin/bitcoin#25797.

hebasto commented:
## Autotools -- CMake Feature Parity Table

Autotool-based build system (AT) features being listed according `./configure --help` output. 

| AT feature | CM feature |
|---|---|
| `--enable-c++20` | TBD |
| `--disable-wallet` | `-DENABLE_WALLET` |
| `--enable-usdt` | `-DWITH_USDT` |
| `--enable-upnp-default` | `-DENABLE_UPNP_DEFAULT` |
| `--enable-natpmp-default` | `-DENABLE_NATPMP_DEFAULT` |
| `--disable-tests` | TBD |
| `--disable-gui-tests` | TBD |
| `--disable-bench` | TBD |
| `--enable-extended-functional-tests` | TBD |
| `--enable-fuzz` | TBD |
| `--enable-fuzz-binary` | TBD |
| `--disable-hardening` | `-DCMAKE_BUILD_TYPE=RelHardened` |
| `--enable-reduce-exports` | `-DREDUCE_EXPORTS` |
| `--disable-ccache` | `-DUSE_CCACHE` |
| `--enable-suppress-external-warnings` | TBD |
| `--enable-lcov` | TBD |
| `--enable-lcov-branch-coverage` | TBD |
| `--enable-threadlocal` | TBD |
| `--disable-asm` | TBD |
| `--disable-zmq` | `-DWITH_ZMQ` |
| `--enable-multiprocess` | TBD |
| `--disable-man` | TBD |
| `--enable-debug` | TBD |
| `--enable-gprof` | TBD |
| `--enable-werror` | TBD |
| `--enable-external-signer` | `-DWITH_EXTERNAL_SIGNER` |
| `--enable-lto` | TBD |
| `--enable-util-cli` | `-DBUILD_CLI` |
| `--enable-util-tx` | `-DBUILD_TX` |
| `--enable-util-wallet` | `-DBUILD_WALLET_TOOL` |
| `--enable-util-util` | `-DBUILD_UTIL` |
| `--enable-experimental-util-chainstate` | TBD |
| `--with-seccomp` | `-DWITH_SECCOMP` |
| `--with-sqlite` | `-DWITH_SQLITE` |
| `--without-bdb` | `-DWITH_BDB` |
| `--with-miniupnpc` | `-DWITH_MINIUPNPC` |
| `--with-natpmp` | `-DWITH_NATPMP` |

...

Rspigler commented:
Is it true that `--without-bdb` is `-DWITH_BDB`?
Could you add some more with/without enable/disable values? For example, what is CM version of `--disable-external-signer`?

hebasto commented:
> Is it true that `--without-bdb` is `-DWITH_BDB`? Could you add some more with/without enable/disable values? For example, what is CM version of `--disable-external-signer`?

More detailed:
- `--with-bdb=auto` or nothing == `-DWITH_BDB=auto` or nothing
- `--with-bdb` or `--with-bdb=yes`  == `-DWITH_BDB=yes`
- `--with-bdb=no` or `--without-bdb`  == `-DWITH_BDB=no`

hebasto commented:
> what is CM version of `--disable-external-signer`?

`-DWITH_EXTERNAL_SIGNER=auto,yes,no`

hebasto commented:
@Rspigler 

[`cmake-gui`](https://cmake.org/cmake/help/latest/manual/cmake-gui.1.html) can be used to see all available options in a nice way.

Rspigler commented:
Thanks!

fanquake commented:
Maybe close this for now, to avoid any confusion, and given we've got the parent PR open on the main repo?

hebasto commented:
> Maybe close this for now, to avoid any confusion, and given we've got the parent PR open on the main repo?

Yes :)
